lib/posix-event/eventfd.c

```diff
@@ -0,0 +1,423 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <vfscore/eventpoll.h>
+#include <vfscore/fs.h>
+#include <vfscore/file.h>
+#include <vfscore/dentry.h>
+#include <vfscore/vnode.h>
+#include <vfscore/mount.h>
+#include <uk/syscall.h>
+#include <uk/essentials.h>
+#include <uk/print.h>
+#include <uk/wait.h>
+#include <uk/mutex.h>
+#include <uk/list.h>
+#include <uk/config.h>
+
+#include <sys/eventfd.h>
+#include <inttypes.h>
+#include <errno.h>
+
+struct eventfd {
+	/** Current value of this eventfd */
+	uint64_t val;
+	/** Indicates if this fd is a semaphore */
+	unsigned int is_semaphore;
+	/** Lock to synchronize access */
+	struct uk_mutex lock;
+	/** Wait queue for blocked writers (i.e., value would overflow) */
+	struct uk_waitq w_wq;
+	/** Wait queue for blocked readers (i.e., value is zero) */
+	struct uk_waitq r_wq;
+	/** List of registered eventpolls. */
+	struct uk_list_head ep_list;
+};
+
+static uint64_t e_inode;
+static struct uk_mutex eventfd_global_lock =
+	UK_MUTEX_INITIALIZER(eventfd_global_lock);
+
+static void eventfd_init(struct eventfd *efd, unsigned int initval, int flags)
+{
+	efd->val = initval;
+	efd->is_semaphore = (flags & EFD_SEMAPHORE);
+	uk_waitq_init(&efd->w_wq);
+	uk_waitq_init(&efd->r_wq);
+	uk_mutex_init(&efd->lock);
+	UK_INIT_LIST_HEAD(&efd->ep_list);
+}
+
+static unsigned int eventfd_events(struct eventfd *efd)
+{
+	unsigned int events = 0;
+
+	if (efd->val > 0)
+		events |= EPOLLIN;
+	if (efd->val < (uint64_t)-1)
+		events |= EPOLLOUT;
+	if (efd->val == (uint64_t)-1)
+		events |= EPOLLERR;
```

Comment:
Here too. The must be another, more elegant, way to write this.

Reply:
Replaced the `(uint64_t)-1` with `UINT64_MAX`, ok?

Reply:
Perfect, thanks a lot :D